### PR TITLE
Show valid values of an enum option in help

### LINKED ISSUE #18: Show valid values of an enum option in help

### PR DESCRIPTION
**Pull request by domn1995**
Closes #18.

**Notable Changes**

- Added extra logic to `HelpTextWriter` that retrieves possible enum values and appends them to the option description.
- Created a functional test validating the feature.
- Created a new dummy class to demo the feature.

**Additional Comment**

I'd also like to bring attention to the following null check:
https://github.com/domn1995/CliFx/blob/7ff83c42066865cb4cf0354ed93694114128810a/CliFx/Domain/HelpTextWriter.cs#L264-L266

This was required even though nullable reference types reported that `Property` was non-nullable. That is due to the damn it operators here:

https://github.com/Tyrrrz/CliFx/blob/cbb72b16ae2ba095abcd593cbddf6e8961ba10b8/CliFx/Domain/CommandOptionSchema.cs#L102-L106

Maybe we can open up an issue to remove those `null!`s and refactor accordingly just as general quality control.

### CODE REVIEW

File: CliFx.Tests.Dummy/Commands/ShowValidEnumValuesCommand.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public class ShowValidEnumValuesCommand : ICommand
+    {
+        [CommandOption("value", Description = "Enum option.", IsRequired = true)]
+        public Values Value { get; set; } = Values.Value1;
+
+        [CommandOption("empty", Description = "Empty enum option.")]
+        public EmptyEnum Empty { get; set; }
+
+        public ValueTask ExecuteAsync(IConsole console) => default;
+    }
```
Comment (Tyrrrz) [inline CliFx.Tests.Dummy/Commands/ShowValidEnumValuesCommand.cs:27]: I don't think we need this. `CliFx.Tests.Dummy` is used for e2e tests for some things that can't be tested via `CliApplication` (e.g. environment variables). If this command is not called in tests, it can be removed. You can use `CliFx.Demo` if you want to showcase something but it's not required.

File: CliFx.Tests/HelpTextSpecs.Commands.cs [hunk 2]
```diff
@@ -82,6 +82,22 @@ private class RequiredOptionsCommand : ICommand
             public ValueTask ExecuteAsync(IConsole console) => default;
         }
 
+        [Command("cmd-with-enum-opts")]
+        private class EnumOptionsCommand : ICommand
+        {
+            public enum ValuesEnum { Value1, Value2, Value3 };
+
+            public enum EmptyEnum { };
```
Comment (Tyrrrz) [inline CliFx.Tests/HelpTextSpecs.Commands.cs:90]: I would say we don't need to test for empty enum. I don't expect that to be a use case at all and even if it is, we can just assume it will work.

File: CliFx/Domain/HelpTextWriter.cs [hunk 3]
```diff
@@ -246,21 +246,45 @@ void RenderOptions()
...
+                    var propertyType = option.Property?.PropertyType;
+
+                    if (propertyType is object && propertyType.IsEnum)
                     {
-                        Render($"(Environment variable: {option.EnvironmentVariableName}).");
-                        Render(" ");
+                        // Cast to object so that we can use linq methods on the Array.
+                        var values = Enum.GetValues(propertyType)
+                            .Cast<object>()
+                            .Select(v => v.ToString().ToLowerInvariant());
```
Comment (Tyrrrz) [inline CliFx/Domain/HelpTextWriter.cs:271]: I think it's better not to do `ToLowerInvariant()` in case the enum fields consist of more than one word (e.g. `Validation.ThrowOnNonZeroExitCode` would turn into `throwonnonzeroexitcode` which is just unreadable).

File: CliFx/Domain/HelpTextWriter.cs [hunk 4]
```diff
@@ -246,21 +246,45 @@ void RenderOptions()
...
+
+                        // Only render valid values if the enum is not empty.
+                        var validValues = values.Any()
+                            ? $" Valid values: {string.Join(", ", values)}."
+                            : "";
```
Comment (Tyrrrz) [inline CliFx/Domain/HelpTextWriter.cs:276]: Again, I think we can just safely assume that the enum has values. I can't imagine a use case where an enum without values is used as option.

File: CliFx/Domain/HelpTextWriter.cs (hunk 3 above)
Comment (Tyrrrz) [inline CliFx/Domain/HelpTextWriter.cs:271]: I would move this logic to `CommandOptionSchema` (or even `CommandArgumentSchema`?) as something like `GetValidValues()`. We will probably also want to support other types in the future. Also, we will likely want this working with `Nullable<Enum>` as well.

File: CliFx/Domain/HelpTextWriter.cs [hunk 5]
```diff
@@ -246,21 +246,45 @@ void RenderOptions()
...
+                    RenderOptionDescription(option);                    
+
+                    RenderNewLine();
+                }
+            }
+
+            void RenderOptionDescription(CommandOptionSchema option)
+            {
+                // Description
+                if (!string.IsNullOrWhiteSpace(option.Description))
```
Comment (Tyrrrz) [inline CliFx/Domain/HelpTextWriter.cs:258]: Looks like valid values and env variable name is only rendered if the description is set. I don't think that's right.

**Comment by Tyrrrz** [comment]
> Maybe we can open up an issue to remove those null!s and refactor accordingly just as general quality control.

I completely agree. Built-in options are a bit of hack right now because they're separate from the actual command. I'm still not sure how to handle it better.

File: CliFx.Tests.Dummy/Commands/ShowValidEnumValuesCommand.cs (hunk 1 above)
Comment (domn1995) [reply CliFx.Tests.Dummy/Commands/ShowValidEnumValuesCommand.cs]: Removed.

File: CliFx.Tests/HelpTextSpecs.Commands.cs (hunk 2 above)
Comment (domn1995) [reply CliFx.Tests/HelpTextSpecs.Commands.cs]: Removed FTs for empty enums.

File: CliFx/Domain/HelpTextWriter.cs (hunk 3 above)
Comment (domn1995) [reply CliFx/Domain/HelpTextWriter.cs]: Moved to `CommandArgumentSchema` and removed call to `ToLowerInvariant()`.

File: CliFx/Domain/HelpTextWriter.cs (hunk 4 above)
Comment (domn1995) [reply CliFx/Domain/HelpTextWriter.cs]: In order to support getting the valid values of other types in the future, I left a similar check in order to not print `Valid Values: ` for unbounded valid values.

File: CliFx/Domain/HelpTextWriter.cs (hunk 3 above)
Comment (domn1995) [reply CliFx/Domain/HelpTextWriter.cs]: Moved and added support for Nullable value types.

File: CliFx/Domain/HelpTextWriter.cs (hunk 5 above)
Comment (domn1995) [reply CliFx/Domain/HelpTextWriter.cs]: Fixed. I had accidentally moved it in the `RenderOptionDescription()` method before which has been removed as of this latest commit.

**Comment by domn1995** [comment]
@Tyrrrz, ready for round two :)

File: CliFx/Domain/CommandArgumentSchema.cs [hunk 6]
```diff
@@ -18,6 +18,37 @@ internal abstract partial class CommandArgumentSchema
...
+            var underlyingType = Nullable.GetUnderlyingType(propertyType);
+            if (underlyingType is object && propertyType.IsValueType)
+            {
+                yield return "null";
+                propertyType = underlyingType;
+            }
+
+            if (propertyType.IsEnum)
+            {
+                // Cast to object so that we can use linq methods on the Array.
```
Comment (Tyrrrz) [inline CliFx/Domain/CommandArgumentSchema.cs]: Looks like the comment is outdated.

File: CliFx/Domain/CommandArgumentSchema.cs [hunk 7]
```diff
@@ -18,6 +18,37 @@ internal abstract partial class CommandArgumentSchema
...
+            {
+                yield break;
+            }
+
+            // Handle nullable values.
+            // If the property is nullable, this will return a non-null value.
+            var underlyingType = Nullable.GetUnderlyingType(propertyType);
+            if (underlyingType is object && propertyType.IsValueType)
+            {
+                yield return "null";
```
Comment (Tyrrrz) [inline CliFx/Domain/CommandArgumentSchema.cs]: I think we don't need to return `null` as valid value, because the user can't actually specify `null`, they instead just not specify the option instead. So let's just have the same behavior for nullable and non-nullable enums where it just returns the list of fields.

File: CliFx/Domain/CommandArgumentSchema.cs [hunk 8]
```diff
@@ -18,6 +18,37 @@ internal abstract partial class CommandArgumentSchema
 
         public bool IsScalar => TryGetEnumerableArgumentUnderlyingType() == null;
 
+        public IEnumerable<string> GetValidValues()
```
Comment (Tyrrrz) [inline CliFx/Domain/CommandArgumentSchema.cs:21]: Let's change the return type to `IReadOnlyList<string>` since we want it buffered and we'll be enumerating it multiple times.

File: CliFx/Domain/CommandArgumentSchema.cs [hunk 9]
```diff
@@ -52,17 +83,17 @@ protected CommandArgumentSchema(PropertyInfo property, string? description)
                 // String-constructable
                 var stringConstructor = GetStringConstructor(targetType);
                 if (stringConstructor != null)
-                    return stringConstructor.Invoke(new object[] {value!});
+                    return stringConstructor.Invoke(new object[] { value! });
```
Comment (Tyrrrz) [inline CliFx/Domain/CommandArgumentSchema.cs:92]: Whitespace changes 😱

File: CliFx/Domain/HelpTextWriter.cs (hunk 4 above)
Comment (Tyrrrz) [reply CliFx/Domain/HelpTextWriter.cs]: Good point

File: CliFx/Domain/CommandArgumentSchema.cs (hunk 9 above)
Comment (domn1995) [reply CliFx/Domain/CommandArgumentSchema.cs:92]: Ooops! I think Roslynator is trolling us whenever I fix the whitespace formatting. Fixed :)

File: CliFx/Domain/CommandArgumentSchema.cs (hunk 6 above)
Comment (domn1995) [reply CliFx/Domain/CommandArgumentSchema.cs]: Fixed.

File: CliFx/Domain/CommandArgumentSchema.cs (hunk 7 above)
Comment (domn1995) [reply CliFx/Domain/CommandArgumentSchema.cs]: Fixed.

File: CliFx/Domain/CommandArgumentSchema.cs (hunk 8 above)
Comment (domn1995) [reply CliFx/Domain/CommandArgumentSchema.cs:21]: Moved this logic a new method `IEnumerable<string> EnumerateValidValues()` and refactoring `GetValidValues()` to return `IReadOnlyList<string>` by lazily caching the result of `EnumerateValidValues()` in a private field.

**Comment by domn1995** [comment]
@Tyrrrz, ready for round three 🤞

**Comment by Tyrrrz** [comment]
Thanks! 🎉
